### ISSUE TITLE
median_iter() has slow performance for higher numbers of chunks

### ISSUE THREAD

**Issue by psavery** [HEXRD/hexrd#306]
[Attached is an example](https://github.com/HEXRD/hexrd/files/6928055/median_time_test.zip), for which you need to download the `RUBY_4537.ge` and `RUBY_4538.ge` files from [here](https://drive.google.com/drive/u/0/folders/1CgGzNEFeaB06I-Q3tbB0lf6nVOjN93fr) to run it.

For example, on my computer, I just ran the script and received the following output:
```bash
Running median()...
median() took 12.79 seconds
Running median_iter() with nchunks=1...
median_iter() with nchunks=1 took 13.02 seconds
Running median_iter() with nchunks=60...
median_iter() with nchunks=60 took 72.32 seconds
```

You can see that that with `nchunks=60`, it took ~5.5x as long to run it compared to `nchunks=1` (and a Windows user reported it taking ~5 minutes for the same data). Since we are using `median_iter()` for progress updates, `nchunks=60` is a reasonable number for us to use in the GUI as well, because it allows us to show an update to the progress bar every 1 or 2 percent.

I believe that the issue is that each frame in the image series is getting expanded for each chunk, which, for our example above, leads to 120 frames * 60 chunks = 7200 frame expansions. This is because each frame is being expanded [here](https://github.com/HEXRD/hexrd/blob/2fef12289aba6adc5212505d83972fd92f2a5b23/hexrd/imageseries/stats.py#L182-L183), and the `_toarray()` function is being called for each chunk [here](https://github.com/HEXRD/hexrd/blob/2fef12289aba6adc5212505d83972fd92f2a5b23/hexrd/imageseries/stats.py#L135).

I understand that part of the issue is that to obtain every median value `ij` in the resulting image, you need `ij` of every frame, and currently, to obtain any single value from a frame, you must expand the whole frame.

If I understand the code correctly, I believe `median()` (which calls [`percentile()`](https://github.com/HEXRD/hexrd/blob/2fef12289aba6adc5212505d83972fd92f2a5b23/hexrd/imageseries/stats.py#L120-L123)) expands the entire image series and keeps it in memory for the `percentile()` call. Perhaps one solution for the performance issue with `percentile_iter()` is to keep in memory as many frames as it can, to reduce the number of times each frame needs to be expanded. It may be able to keep in memory all of the frames if there is enough memory on the system, and thus match the performance of `median()` / `percentile()`.

**Comment by donald-e-boyce** [comment]
I'm working on this.

**Comment by psavery** [comment]
Hi Don! Here are my timing results for my test data:

Master branch:
```
median() took 12.65 seconds
median_iter() with nchunks=1 took 12.16 seconds
median_iter() with nchunks=60 took 65.77 seconds
```

Your branch:
```
median() took 12.69 seconds
median_iter() with nchunks=1 took 12.42 seconds
median_iter() with nchunks=60 took 8.24 seconds
```

It looks like it is much faster than it was before, which is great! I don't immediately know why more chunks would be faster.

**Comment by donald-e-boyce** [comment]
Wow! That,s great. 

I added a keyword arg to percentile_iter and median_iter, use_buffer, True by default. In my tests above, I thought I ran it with both True and False, but there was a bug in my test and I didn't. All makes sense now. Here are new results with 50 frames.

    ===== Using 50 frames =====
    stats.median:  8.848058950999999
    
    stats._toarray:  0.9237148639999972
    np.median:  7.592521657000006
    
    nchunks =  1  buffer on
    stats.median_iter:  8.814753901000003
       chunks= 1
       frames= 50
       difference:  0.0
    nchunks =  1  buffer off
    stats.median_iter:  8.499511976999997
       chunks= 1
       frames= 50
       difference:  0.0
    
    nchunks =  5  buffer on
    stats.median_iter:  4.027293435999994
       chunks= 5
       frames= 50
       difference:  0.0
    nchunks =  5  buffer off
    stats.median_iter:  6.351542835999993
       chunks= 5
       frames= 50
       difference:  0.0
    
    nchunks =  10  buffer on
    stats.median_iter:  3.950250507000007
       chunks= 10
       frames= 50
       difference:  0.0
    nchunks =  10  buffer off
    stats.median_iter:  9.805906575000009
       chunks= 10
       frames= 50
       difference:  0.0
    
    nchunks =  20  buffer on
    stats.median_iter:  4.039551971999998
       chunks= 20
       frames= 50
       difference:  0.0
    nchunks =  20  buffer off
    stats.median_iter:  16.928544521999996
       chunks= 20
       frames= 50
       difference:  0.0
    
    nchunks =  50  buffer on
    stats.median_iter:  4.245434304000014
       chunks= 50
       frames= 50
       difference:  0.0
    nchunks =  50  buffer off
    stats.median_iter:  37.38503416099999
       chunks= 50
       frames= 50
       difference:  0.0

**Comment by psavery** [comment]
This was fixed by #309
